761-signal-exit

jlitzingerdev opened the pull request:
Pull request for https://twistedmatrix.com/trac/ticket/761

jlitzingerdev committed:
Capture the signal that caused the reactor to exit.

Currently, if a signal causes the reactor to exit, that signal is lost
to the user.  This poses a problem for clients interested in the signal
to ensure a proper exit code.

Capturing the signal in a class attribute allows interested parties
(e.g. twistd) to obtain the value for later use.

jlitzingerdev committed:
Select reactor in startReactor

If the caller hasn't selected a reactor, choose the default now so that
the ApplicationRunner has access to the reactor after
runReactorWithLogging returns.  Without a reference to the reactor, the
application runner has no way of accessing the exitSignal attribute to
properly set the exit status.

The redundancy here (runReactorWithLogging also sets the reactor, so
that could have been lifted) to avoid modifying runReactorWithLogging.
While twisted doesn't currently use it in many places, it is a public
api and may be used by existing client code.

jlitzingerdev committed:
Save the value of the reactor's exitSignal attribute if it exists.

The exitSignal attribute is nonzero on a reactor when the reactor exits
due to a signal (on supported platforms).  When this value is set, save
it in the application runner so that callers (e.g. twistd) can exit with
the appropriate exit status.

jlitzingerdev committed:
Add exitWithSignal to allow exiting with a specific signal.

exitWithSignal() provides a mechanism for the user to deliver a
specified signal to the current process with os.kill.

jlitzingerdev committed:
Ensure twistd exits with a signal if one is caught by the reactor.

If the reactor selected by the ApplicationRunner exits with a signal
twistd will reset the signal handler to the default and deliver the
signal to itself via os.kill.

jlitzingerdev committed:
If twist's runner exits with a signal, deliver the signal to exit.

Currently, if twist's runner exits with a signal that is handled by the
reactor, that signal will not be passed to the shell on exit.  If the
reactor exits with a signal, deliver that signal with os.kill to exit.

jlitzingerdev committed:
Add news fragment for ticket 761.

New topfile containing a brief summary of most visible affects of the
changes, namely that twist and twistd will now use os.kill to redeliver
a signal to themselves should they exit with a signal.

jlitzingerdev commented:
Doh.  That missed coverage was a stupid copy/paste error on my part in the test case.  Fixing.

jlitzingerdev committed:
Ensure test cases invoke the correct method.

The SIGTERM and SIGBREAK tests were not calling the appropriate method,
yet the tests pass because all the handlers end up doing the same thing
under the hood (setting the exitSignal attribute).

jlitzingerdev commented:
At this point, assuming I properly interpreted the results of codecov (still learning the tool), it looks like the only missed lines in the diff are those that were skipped (e.g. the sigBreak path).  If desired I can easily fake that path, but wasn't sure what the appropriate action was considering that signal will/will not exist on all platforms.

glyph commented on src/twisted/application/app.py:
```diff
@@ -681,3 +690,17 @@ def startApplication(application, save):
+
+
+
+def exitWithSignal(sig):
```
I don't think this needs to be public API.  We should be conservative about exporting new public API, preferring to go with private APIs at first and exposing them only deliberately.

glyph commented on src/twisted/application/app.py:
```diff
@@ -681,3 +690,17 @@ def startApplication(application, save):
+    @type sig:  C{int}
+
+    """
+    signal.signal(sig, signal.SIG_DFL)
```
Rather than mutating global signal state, can't this just be accomplished with `os._exit(0b10000000 | sig)`?

jlitzingerdev replied:
@glyph  Yes but I don't think that is a valid thing to do in all cases?.  If setting the exit status to 128 + signal was always the right thing to do on all platforms, this would (I think) work.  I actually went this route initially, but changed after reading some of the referenced articles in the original ticket.  My concern is that, while the GNU C library certainly obeys the above behavior, I haven't found any spec/guarantee to indicate this would be consistent with other implementations.  Thus, rather than assume it is ok, I chose to let the OS do what it would normally do had the signal not been caught in the first place.  Comment 8 in the ticket, and the linked article talk a little about calling exit.

In approaching this bug I looked for some consistent, authoritive reference on valid exit status', and realized it is somewhat of a platform/implementation specific mess.  Therefore, lacking a better alternative, I wanted to let the OS take control and do what it would normally do.

exarkun replied:
> My concern is that, while the GNU C library certainly obeys the above behavior, I haven't found any spec/guarantee to indicate this would be consistent with other implementations. 

Correct.  There is no POSIX API for _constructing_ an exit code - only deconstructing it.  You probe the system to create some  values and then inspect them to see how the system composes the two pieces of information and then compose them yourself that way ... but ...  you could also just re-deliver the signal. :)

glyph commented on src/twisted/application/twist/test/test_twist.py:
```diff
@@ -201,3 +201,102 @@ def run(self):
+
+
+
+class CrippledRunner(object):
```
1. "Crippled" could be construed as insulting to disabled people, so please don't use the term in the codebase.
2. It's also not a standard technical term; I think you mean "dummy"? http://xunitpatterns.com/Mocks,%20Fakes,%20Stubs%20and%20Dummies.html

jlitzingerdev replied:
No problem here.  Btw, I used it in a similar context to where it is used elsewhere in the code (see test_twistd).

glyph commented on src/twisted/application/twist/test/test_twist.py:
```diff
@@ -201,3 +201,102 @@ def run(self):
+                return reactor
+
+
+        self.patch(_twist, "TwistOptions", TestTwistdOptions)
```
This level of patching suggests that this is not a good testing strategy :-\.

jlitzingerdev replied:
Fair,  I have an alternate idea, but I'll raise questions in the ticket.

glyph commented on src/twisted/application/twist/test/test_twist.py:
```diff
@@ -201,3 +201,102 @@ def run(self):
+            def installReactor(self, name):
+                reactor = MemoryReactor()
+                return reactor
+
```
(nitpick: too many blank lines. I wonder why twistedchecker isn't complaining?)

jlitzingerdev replied:
It does, I just missed it, will fix in v2.

glyph commented on src/twisted/internet/base.py:
```diff
@@ -473,6 +473,8 @@ class ReactorBase(object):
         register the thread it is running in as the I/O thread when it starts.
         If C{True}, registration will be done, otherwise it will not be.
+
+    @ivar exitSignal: The signal that cause the reactor to stop, or 0.
```
This attribute appears on some implementations of the reactor now but not others, it's never documented on an interface.

jlitzingerdev replied:
This is one of the bigger challenges I had, will discuss in the ticket.

glyph reviewed (commented):
Thanks for submitting this fix!  It mostly looks good, but I do have some concerns about exposing a bunch of new public API as a result.  Could you fix this to be an implementation detail somehow without exposing it?

exarkun commented on src/twisted/application/twist/test/test_twist.py:
```diff
@@ -201,3 +201,102 @@ def run(self):
+        def startService(service):
+            pass
+
+        self.patch(MultiService, "startService", startService)
```
This looks kind of sad.  What's going on?  I suggest refactoring something so you don't have to do this.

jlitzingerdev replied:
Thanks!
The intent was to let Twist invoke the plugin/service machinery normally, but prevent the service from actually starting.  I took the patching approach because it was already used in the test module to achieve the same goal (as well as additional goals specific to that test).  I'm fairly certain it isn't the only way to test these changes without running the service so yeah, agree, I'll refactor.

jlitzingerdev replied:
Actually, changing the plugin called (e.g. to words) would also eliminate the need for that patching as well.  When the web plugin's makeService executes it calls strports.service, which imports the default reactor if one isn't passed in and results in the reactory being dirty when the test exits.

Regardless, I'd prefer something a little more deterministic than relying on the current implementation of whatever plugin is chosen for the test.

jlitzingerdev committed:
Merge branch 'trunk' into 761-signal-exit

Manually resolve conflicts.  Recent changes eliminated the
runnerArguments parameter in favor of an options class.  Refactor the
changes in this branch to account.

jlitzingerdev committed:
Make twisted.application.app.exitWithSignal private.

Incorporatei review comments to make exitWithSignal a private function.

jlitzingerdev committed:
Make _exitSignal class attribute private on ReactorBase.

A concern raised during review is that this attribute exists on some
reactors, but not others, but isn't documented in any interface.  This
characteristic is true of signal handling by reactor implementations --
only some inherit _SignalReactorMixin and handle signals.  Since this
attribute is coupled to the same functionality, its visibility should
match that of signal handling.

jlitzingerdev committed:
Handle exit within Twisted.run.

If when Runner.run returns, determine whether a signal caused the
reactor to stop and deliver appropriately.  Moving this logic into run
allows for better testing (less patching) without trading anything in
return.

In testing, patch as little as possible, specifically, the
_exitWithSignal and startLogging methods must be patched.  The latter
because even delivering a dummy factory will result in a new observer
that will cause other, unrelated tests to fail.

jlitzingerdev committed:
Fix 'open/close' docstring errors.

Fix instances of 'the opening/closing of docstring should be
on a line by themselves' introduced by this patch series.

jlitzingerdev committed:
Modify test class name to match xunit conventions.

The SignalCapturing reactor is being used to verify the outputs of
running the reactor and delivering it a signal.  Rather than an
arbitrary name, prefix with a common term taken from xunit.

http://xunitpatterns.com/Test%20Spy.html

jlitzingerdev committed:
Fix missing docstring.

The test class was missing a class docstring and added a new
twistedchecker warning.

jlitzingerdev committed:
Fix blank lines style check warnings.

Twistedchecker was raising errors due to the number of blank lines in
various contexts being wrong.

jlitzingerdev committed:
Add missing docstrings.

Test class was missing docstrings for itself and setUp.

jlitzingerdev committed:
Fix variable names to match twisted convention.

twistedchecker was flagging these variables because they do not conform
to twisted's naming convention.

jlitzingerdev committed:
Eliminate code not relevant to the test.

Neither the called variable or the RuntimeError are relevant to what is
being tested.  The purpose of the DummyReactorWithSignal is to provide
an _exitSignal attribute to the ApplicationRunner.

jlitzingerdev committed:
Merge branch 'trunk' into 761-signal-exit

glyph commented on src/twisted/application/app.py:
```diff
@@ -392,8 +395,14 @@ def startReactor(self, reactor, oldstdout, oldstderr):
             self.config, oldstdout, oldstderr, self.profiler, reactor)
+        try:
+            self._exitSignal = reactor._exitSignal
+        except AttributeError:
```
I think we should be _extremely_ cagey about `AttributeError`s, especially in the presence of an `except: pass`.

For example, imagine a future scenario in which someone adds `__slots__` to ApplicationRunner, but forgets to add this attribute.  Of course, in *this* case, a test would nicely fail and explain what had gone wrong, but if someone cribs from this example they'll produce this error elsewhere.

So I'd rewrite this

```python
try:
    exitSignal = reactor._exitSignal
except:
    exitSignal = None
self._exitSignal = exitSignal
```

and then you can also eliminate the redundant cvar.

glyph replied:
Furthermore…

Given that `_exitSignal` is a private attribute, it's always possible that some 3rd-party reactor is using it.

As such, an unambiguous way to signal [namespace ownership](https://glyph.twistedmatrix.com/2017/05/the-sororicide-antipattern.html) would be helpful here.  Interfaces can do this.

Basically, you make a private interface, `_ISupportsExitSignaling`, make every reactor in Twisted that uses this functionality provide it, and then here you can do `if _ISupportsExitSignaling.providedBy(reactor)` before accessing this attribute.  This is unambiguous, and makes it easier for individual reactors to document what they're doing when they provide this attribute.  How does this sound?

jlitzingerdev replied:
I like this, it's much clearer, so yes, I'm for this implementation.

glyph commented on src/twisted/application/app.py:
```diff
@@ -681,3 +690,17 @@ def startApplication(application, save):
+
+    @param sig:  Signal to use to terminate the process with C{os.kill}.
+    @type sig:  C{int}
+
```
Just as a stylistic note: this blank line really shouldn't be here.

glyph commented on src/twisted/application/twist/_twist.py:
```diff
@@ -99,6 +100,12 @@ def run(twistOptions):
+            if reactor._exitSignal:
+                _exitWithSignal(reactor._exitSignal)
+        except:
+            pass
```
`except: pass` is pretty much always wrong; it makes debugging impossible.  What exceptions are you actually expecting here?  Why is it important to catch them?

jlitzingerdev replied:
In general, I agree, and I really didn't like it when I added it, though I did have a purpose in mind.  I'll explain why, and if you feel it isn't necessary I'm more than happy to remove it.

The purpose here was to preserve the existing exit semantics of twist in the event _exitWithSignal throws an exception.  While that shouldn't happen (technically, I'm not confident that os.kill wont ever throw on other platforms, I'm primarily familiar with POSIX and Linux in particular), my inability to fully verify it can't happen led me to add the catch-all.

Specifically, if an exception gets thrown, I assume it'll go up to the interpreter, and from there may or may not affect the exit status of the program.  Rather than have this new code trigger that, I caught the exception and let the program exit as it does today.

glyph commented on src/twisted/application/twist/test/test_twist.py:
```diff
@@ -191,3 +191,80 @@ def run(self):
+            """
+            pass
+
+        self.patch(Runner, 'startLogging', startLogging)
```
`patch` on a global object is always a form of technical debt.

Is there any possible way you can inject a `Runner` instance into `Twist`?  This could be a follow-up if you don't want to set up this plumbing in this PR, I just don't want to lose track of it.

jlitzingerdev replied:
If memory serves, we don't even have to inject a dummy instance.  I did an interation with a Dummy log factory and the tests will work just fine.  The problem was that downstream logging specific tests failed.  I inferred, from a quick glance at the failure, that those tests were expecting only a certain number of global observers.  We can likely avoid that by manually mutating the global observers if that is possible.  Let me redo it that way and see exactly what those errors were.

jlitzingerdev replied:
Specifically, one of the errors when a dummy logger instance is passed is:
```
twisted.trial.unittest.FailTest: 4 != 1

twisted.test.test_log.LogTests.test_erroneousErrors
```
I'll look into what the failing test are expecting, but I suspect removing the created observers in my test will do the trick.

jlitzingerdev replied:
Actually, no, that's not at all what is happening.  I'll look at either using the existing runner or injecting one.

jlitzingerdev replied:
@glyph I've looked in to this a bit, and there aren't many good options with the existing code.  Runner is imported into _twist, it isn't an option that can be easily injected.  The default runner can be used, but somewhere the call to `globalLogBegginer.beginLoggingTo` needs to be prevented or downstream tests fail.

WIth the existing code, the following seem like the options:
1.  Leave it as is.
2.  Patch the Runner itself, as is done in other tests.
3.  Patch globalLogBeginner in twisted.application.runner._runner with a new LogBeginner for each test.
4.  Patch 'beginLoggingTo' on globalLogBeginner as is done in test_twistd.

Options 3 and 4 allow the use of the default runner.

To inject a runner, the runner should probably be sourced from TwistOptions, rather than imported globally, and be selectable.  While I am more than willing to do this change, it does feel like it is significant enough to warrant its own ticket.  Specfiically, it seems like a fairly heavy side effect for an unrelated change.

Thoughts?

glyph commented on src/twisted/internet/test/test_base.py:
```diff
@@ -369,3 +369,65 @@ def test_str(self):
+        """
+        Required method, unused.
+        """
+        pass
```
`pass` is not necessary here, the docstring is sufficient.  (Sorry, twistedchecker ought to check for this.)

jlitzingerdev replied:
I was thinking about trying to add the ability to check a patchfile (similar to the Linux kernel's checkpatch utility) to twistedchecker.  The benefit there being that it would potentially ignore previous errors in other unrelated parts of the file.  If I go down that road I'll look at adding this too if nobody beats me to it. 

Will correct in V3.

glyph commented on src/twisted/test/test_twistd.py:
```diff
@@ -1911,5 +1956,83 @@ def test_hooksNotCalled(self):
 
 
+class ExitWithSignalTests(unittest.TestCase):
+
```
There should not be a blank line here.

jlitzingerdev replied:
Should twistedchecker flag this?  It didn't for me (though I did miss another "too many blank lines" warning later in the class.

glyph commented on src/twisted/test/test_twistd.py:
```diff
@@ -1911,5 +1956,83 @@ def test_hooksNotCalled(self):
+
+
+    def tearDown(self):
+        CrippledApplicationRunner._exitSignal = 0
```
Is this expected to be set by the system under test?  If not, it would be more appropriate to use `.patch` for this.

jlitzingerdev replied:
The real need is to inject an input into the SUT, so a Stub ApplicationRunner is likely more appropriate.

glyph commented on src/twisted/test/test_twistd.py:
```diff
@@ -1911,5 +1956,83 @@ def test_hooksNotCalled(self):
+        _exitWithSignal is not called if the runner does not exit with a
+        signal.
+        """
+        self.patch(twistd.app, '_exitWithSignal', self.fakeExitWithSignal)
```
This is still an awful lot of patching.  Is there any way you can think of to reduce it in the future?

jlitzingerdev replied:
I didn't make heavy changes in v2 to the test_twistd, so I'm going to have to say, not sure?  The goal here is to make sure _exitWithSignal is called.  Calling the real function would deliver a signal to the test, which would be bad, so it seems like some patching is required with the current implementation.  I could push the patching back to `os.kill()`, which might loosely be considered a platform integration point where patching is more appropriate, thoughts?

Not saying that's the only way, just the immediately obvious one.  I'll take a look.

glyph reviewed (commented):
I think this is definitely approaching mergeable, but I'd like you to at least consider the remaining patches and the `_exitSignal` namespace confusion issue.  If you think it would be too much work to address those we can probably merge it as-is with some follow-up tickets.

jlitzingerdev committed:
Merge branch 'trunk' into 761-signal-exit

jlitzingerdev committed:
Introduce _ISupportsExitSignalCapturing interface.

The new interface  is a behavior whereby implementors commit to exposing
the value of any exit (e.g. SIGTERM, SIGINT, SIGBREAK) signal delivered to
a registered handler in the _exitSignal attribute.

jlitzingerdev committed:
ReactorBase implements _ISupportsExitSignalCapturing

By implementing this interface, ReactorBase commits to setting
_exitSignal with the value of any captured exit signal.

There is an alternative way to implement this, in the reactor subclasses
themselves, so it's worth documenting why I chose to implement the
interface in ReactorBase.  A brief rationale follows.

The signals referenced in this commit are exit signals, i.e. those
commonly used to exit a program such as SIGTERM, SIGINT, and SIGBREAK.
Specifically, this does not refer to SIGCHILD.

Currently, two primary actors are involved in signal handling by
reactors -- _SignalReactorMixin and ReactorBase.  The current pattern
is to inherit the mixin at the specific reactor (e.g. PosixReactorBase),
relying on the fact that _SignalReactorMixin is designed to work with a
subclass of ReactorBase.  Specifically, it expects the sigInt, sigTerm,
and sigBreak methods to exist (these methods are not part of any
interface).

This means that while the decision to install the handlers is made by
inheriting _SignalReactorMixin, the implementation of the handler lives
in ReactorBase.  Thus, rather than require each individual reactor to
implement their own (possibly delegating) sigInt, sigTerm, etc.,
implementing _ISupportsExitSignalCapturing in ReactorBase keeps the
responsibility for capturing the integer signal with the code that
handles the signal.

jlitzingerdev committed:
Check whether reactor implements _ISupportsExitSignalCapturing

Rather than catch an AttributeError if the reactor used by
ApplicationRunner, check to see if the reactor implements
_ISupportsExitSignalCapturing reactor.  If not, set an instance
attribute _exitSignal to None.

jlitzingerdev committed:
Modify check of _exitSignal against None

The existing check is valid, however the semantics for _exitSignal are
an integer or None, so an explicit test against None is a clearer
check for the interesting condition, whether or not _exitSignal has been
set to a signal value.

jlitzingerdev committed:
Refactor ExitWithSignalTests to push patching back to os.

The existing code patched the function that was being tested, and
use a test stub to inject the signal value into the system under test,
rather than setting the attribute manually.

jlitzingerdev committed:
Eliminate unecessary pass.

Caught in review, the pass is not necessary.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Fix apidocs build error.

There was an extra colon, which broke the apidocs build.

jlitzingerdev committed:
Resolve apidocs reference warnings.

The leading _ was missing from the interface reference.

jlitzingerdev committed:
Check whether reactor implements _ISupportsExitSignalCapturing

Rather than the catch-all try/except, check whether the reactor used by
twist implements _ISupportsExitSignalCapturing and, if it is None,
exiting with the captured signal.

jlitzingerdev committed:
Eliminate line after class name.

This line should not be here, though it was not caught by twistedchecker
as of this commit.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Remove blank line.

This was caught during review, remove the blank line at the end of the
docstring.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

wsanchez committed:
Merge branch 'trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

markrwilliams committed:
Merge branch 'trunk' into 761-signal-exit

jlitzingerdev commented:
@markrwilliams @rodrigc Is the Appveyor failure something related to my PR specifially?  If I read it correctly it is complaining about an openssh install, so it doesn't look like it, but if there's something I should look at let me know, otherwise I'll try to keep merging every day or so.

rodrigc commented:
See:
 https://twistedmatrix.com/pipermail/twisted-python/2017-August/031596.html

jlitzingerdev commented:
Thanks, I'll keep an eye for resolution and keep up to date accordingly.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

Conflicts in src/twisted/test/test_twistd.py resolved manually.  The
conflicts were due to moving of imports in a069c3e822, and with the
exception of removing one line from this branch, both sets of changes
were kept.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

exarkun committed:
Merge branch 'trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Fix twistedchecker warnings.

Two instances of docstring open/close and one spacing between class
method issue.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Fix typo in newsfragment.

The change description stated that prior to the change an exit code was
returned, rather than explicitly stating that a zero exit code was
returned despite exiting due to a signal.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 761-signal-exit

glyph committed:
Merge branch 'trunk' into 761-signal-exit
